File: cfr.php
<?php

class CFR
{
    public String $type;
    public String $id;
    public Int $window;


    public function __construct(String $type, String $id, Int $TimeWindow = 1)
    {
        $this->type = $type;
        $this->id = $id;
        $this->window = $TimeWindow;
    }

    private function parseText(String $text): array
    {
        $lines = explode("\n", $text);
        $lines = array_map(fn ($line) => trim($line, " \r"), $lines);
        $start_line = array_search('//<![CDATA[', $lines) + 1;
        $end_line = array_search('//]]>', $lines) - 1;
        $lines = array_slice($lines, $start_line, $end_line - $start_line + 1);
        $lines = array_filter($lines, fn ($line) => str_starts_with($line, 'VALUES'));
        return $lines;
    }

    private function parseLine(String $line): array
    {
        $stripped = explode(');', explode('Array(', $line, 2)[1], 2)[0];
        $values = explode(',', $stripped, 4);
        $values = array_map(fn ($val) => trim($val, ' "'), $values);
        $date = DateTime::createFromFormat(($this->window==1) ? 'd/m/Y H.i' : 'd/m/Y', $values[1]);

        return [
            'id' => intval($values[0]),
            'timestamp' => $date->getTimestamp(),
            'values' => [
                $values[2],
                $values[3]
            ]
        ];
    }

    public function getData(): array
    {

        $id = $this->id;
        $type = $this->type;
        if ($this->window == 30) {
            $type .= '_men';
        }

        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => "https://www.cfr.toscana.it/monitoraggio/dettaglio.php?id=$id&type=$type",
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => '',
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST => 'GET',
            CURLOPT_HTTPHEADER => array(
                'Cookie: language=IT; speed=m%2Fs'
            ),
        ));
        $response = curl_exec($curl);
        curl_close($curl);

        $lines = $this->parseText($response);

        $parser = fn ($line) => [];
        switch ($this->type) {
            case 'idro':
                $parser = fn ($line) => $this->parseHydroLine($line);
                break;
            
            case 'pluvio':
                $parser = fn ($line) => $this->parsePluvioLine($line);
                break;

            case 'termo':
                $parser = fn ($line) => $this->parseThermoLine($line);
                break;
            
            case 'anemo':
                $parser = fn ($line) => $this->parseAnemoLine($line);
                break;

            case 'igro':
                $parser = fn ($line) => $this->parseHygroLine($line);
                break;
        }


        $values = array_map(fn ($line) => $parser($line), $lines);
        return $values;
    }

    private function parseHydroLine(String $line): array
    {
        $values = $this->parseLine($line);
        return [
            'timestamp' => $values['timestamp'],
            'level' => floatval($values['values'][0])
        ];
    }

    private function parsePluvioLine(String $line): array
    {
        $values = $this->parseLine($line);
        return [
            'timestamp' => $values['timestamp'],
            'level' => floatval($values['values'][0]),
            'cumulative' => floatval($values['values'][1])
        ];
    }

    private function parseThermoLine(String $line): array
    {
        $values = $this->parseLine($line);
        return [
            'timestamp' => $values['timestamp'],
            'temperature' => floatval($values['values'][0])
        ];
    }

    private function parseAnemoLine(String $line): array
    {
        $values = $this->parseLine($line);
        $wind_values = explode('/', $values['values'][0], 2);
        return [
            'timestamp' => $values['timestamp'],
            'speed' => floatval($wind_values[0]),
            'burst' => floatval($wind_values[1]),
            'direction' => floatval($values['values'][1])
        ];
    }

    private function parseHygroLine(String $line): array
    {
        $values = $this->parseLine($line);
        return [
            'timestamp' => $values['timestamp'],
            'humidity' => floatval($values['values'][0])
        ];
    }
}

$cfr = new CFR('pluvio', 'TOS03002016', 1);
$data = $cfr->getData();
print_r($data);

File: cfr.py
from datetime import datetime
import aiohttp
from aiohttp import web
import json
from enum import Enum

PORT = 8547

class CFR:
    class Type(Enum):
        HYDRO   = 'idro'
        PLUVIO  = 'pluvio'
        THERMO  = 'termo'
        ANEMO   = 'anemo'
        HYGRO   = 'igro'

    class TimeWindow(Enum):
        DAILY   = 1
        MONTHLY = 30

    def __init__(self, type:'CFR.Type', station_id:'str', time_window:'TimeWindow'= TimeWindow.DAILY) -> None:
        self.type = type
        self.id = station_id
        self.window = time_window
        self.parser = {
            self.Type.HYDRO: self.__parseHydroLine,
            self.Type.PLUVIO: self.__parsePluvioLine,
            self.Type.THERMO: self.__parseThermoLine,
            self.Type.ANEMO: self.__parseAnemoLine,
            self.Type.HYGRO: self.__parseHygroLine,
        }[self.type]

    def __parseText(self, text:'str') -> 'list[str]':
        lines = [line.strip(' \r') for line in text.split('\n')]
        start_line = lines.index('//<![CDATA[')
        end_line = lines.index('//]]>')
        return [line for line in lines[start_line+1:end_line-1] if line.startswith('VALUES')]

    def __parseLine(self, line:'str') -> 'tuple[int, datetime, str, str]':
        id_str, date_str, value, direction = [value.strip(' "') for value in line.split('Array(',1)[1].split(');',1)[0].split(',',4)]
        strp_format = '%d/%m/%Y %H.%M' if self.window == self.TimeWindow.DAILY else '%d/%m/%Y'
        return int(id_str), datetime.strptime(date_str, strp_format), value, direction

    async def getData(self):
        params = {
            'id':   self.id, 
            'type': self.type.value
        }

        if self.window == self.TimeWindow.MONTHLY:
            params['type'] += '_men'

        async with aiohttp.ClientSession() as session:
            async with session.post("https://cfr.toscana.it/monitoraggio/dettaglio.php", params=params) as resp:
                lines = self.__parseText(await resp.text())
                return [(self.parser(line)) for line in lines] 

    def __parseHydroLine(self, line:'str') -> 'dict':
        _, date, value_str, _ = self.__parseLine(line)
        return {'timestamp':int(date.timestamp()), 'level':float(value_str)}

    def __parsePluvioLine(self, line:'str') -> 'dict':
        _, date, value_str, cumulative_str = self.__parseLine(line)
        return {'timestamp':int(date.timestamp()), 'level':float(value_str), 'cumulative':float(cumulative_str) if len(cumulative_str)>0 else None}

    def __parseThermoLine(self, line:'str') -> 'dict':
        _, date, value_str, _ = self.__parseLine(line)
        return {'timestamp':int(date.timestamp()), 'temperature':float(value_str)}

    def __parseAnemoLine(self, line:'str') -> 'dict':
        _, date, value_str, direction_str = self.__parseLine(line)
        speed, burst = value_str.split('/',1)
        return {'timestamp':int(date.timestamp()), 'speed':float(speed), 'burst':float(burst), 'direction':float(direction_str)}

    def __parseHygroLine(self, line:'str') -> 'dict':
        _, date, value_str, _ = self.__parseLine(line)
        return {'timestamp':int(date.timestamp()), 'humidity':float(value_str)}
        
app = web.Application()
routes = web.RouteTableDef()

@routes.get('/{type}/{station_id}')
async def get_handler(request):
    identifier = str(request.match_info['station_id'])

    type = str(request.match_info['type'] )
    time_window = CFR.TimeWindow.DAILY
    if type.endswith('_men'):
        time_window = CFR.TimeWindow.MONTHLY
        type = type[:-4]
    type = CFR.Type(type)

    js = await CFR(type, identifier, time_window).getData()
    return web.Response(text=json.dumps(js, indent=2), content_type="application/json")

app.add_routes(routes)

web.run_app(app, port=PORT)
